File: app/src/main/java/com/example/spotibae/Adapter/Adapter.java
package com.example.spotibae.Adapter;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.spotibae.Activities.Messaging.MessagingScreen;
import com.example.spotibae.Activities.Messaging.MusicPlayer;
import com.example.spotibae.Activities.User.UserProfile;
import com.example.spotibae.R;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.spotibae.Models.ModelClass;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;

import java.util.List;

public class Adapter extends RecyclerView.Adapter<Adapter.ViewHolder> {

    private List<ModelClass> userList;

    public Adapter(List<ModelClass>userList) {
        this.userList=userList;
    }

    @NonNull
    @Override
    public Adapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view= LayoutInflater.from(parent.getContext()).inflate(R.layout.fragment_chat_user, parent,false);
        ViewHolder holder = new ViewHolder(view, new ViewHolder.MyClickListener() {
            @Override
            public void onMessages(int p) {
                Intent intent = new Intent(view.getContext(), MessagingScreen.class);
                intent.putExtra("PROFILE_PIC_EMAIL", userList.get(p).getEmail());
                intent.putExtra("USER_FIREBASE_ID", userList.get(p).getFirebaseId());
                intent.putExtra("USER_NAME", userList.get(p).getTextview1());
                view.getContext().startActivity(intent);
            }

            @Override
            public void onMusic(int p) {
                Intent intent = new Intent(view.getContext(), MusicPlayer.class);
                intent.putExtra("PROFILE_PIC_EMAIL", userList.get(p).getEmail());
                intent.putExtra("USER_FIREBASE_ID", userList.get(p).getFirebaseId());
                intent.putExtra("USER_NAME", userList.get(p).getTextview1());
                view.getContext().startActivity(intent);
            }
        });
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull Adapter.ViewHolder holder, int position) {
        String firebaseID = userList.get(position).getFirebaseId();
        String email = userList.get(position).getEmail();
        String name=userList.get(position).getTextview1();
        String msg=userList.get(position).getTextview2();
        String time=userList.get(position).getTextview3();
        holder.setData(firebaseID,email,name,msg,time);
    }

    @Override
    public int getItemCount() {
        return userList.size();
    }

    //view holder class

    public static class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        MyClickListener listener;
        public ImageView imageView;
        private TextView textView;
        private TextView textView2;
        private TextView textview3;
        ImageView music;
        ImageView messages;

        public ViewHolder(@NonNull View itemView, MyClickListener listener) {
            super(itemView);
            //here use xml ids
            //give different name not like constructor
            imageView = itemView.findViewById(R.id.imageview);
            textView = itemView.findViewById(R.id.textview);
            textView2 = itemView.findViewById(R.id.textview2);
            textview3 = itemView.findViewById(R.id.textview3);
            music = itemView.findViewById(R.id.musicPlayerButton);
            messages = itemView.findViewById(R.id.messagesButton);

            this.listener = listener;

            music.setOnClickListener(this);
            messages.setOnClickListener(this);
        }

        public void setData(String firebaseId, String email, String name, String msg, String time) {
            //imageView.setImageResource(resource);
            setImage(email);
            textView.setText(name);
            textView2.setText(msg);
            textview3.setText(time);
        }

        @Override
        public void onClick(View view) {
            switch (view.getId()) {
                case R.id.messagesButton:
                    listener.onMessages(this.getLayoutPosition());
                    break;
                case R.id.musicPlayerButton:
                    listener.onMusic(this.getLayoutPosition());
                    break;
                default:
                    break;
            }
        }

        public interface MyClickListener {
            void onMessages(int p);
            void onMusic(int p);
        }

        public void setImage(String email) {
            StorageReference storageReference = FirebaseStorage.getInstance().getReference();
            StorageReference photoReference = storageReference.child("User").child(email).child("profilePic.png");

            final long ONE_MEGABYTE = 1024 * 1024;
            photoReference.getBytes(ONE_MEGABYTE).addOnSuccessListener(new OnSuccessListener<byte[]>() {
                @Override
                public void onSuccess(byte[] bytes) {
                    Bitmap bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
                    imageView.setImageBitmap(Bitmap.createScaledBitmap(bitmap,  90,90, true));
                }
            }).addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception exception) {
                }
            });
        }

        public Bitmap getCroppedBitmap(Bitmap bitmap) {
            Bitmap output = Bitmap.createBitmap(bitmap.getWidth(),
                    bitmap.getHeight(), Bitmap.Config.ARGB_8888);
            Canvas canvas = new Canvas(output);

            final int color = 0xff424242;
            final Paint paint = new Paint();
            final Rect rect = new Rect(0, 0, bitmap.getWidth(), bitmap.getHeight());

            paint.setAntiAlias(true);
            canvas.drawARGB(0, 0, 0, 0);
            paint.setColor(color);
            // canvas.drawRoundRect(rectF, roundPx, roundPx, paint);
            canvas.drawCircle(bitmap.getWidth() / 2, bitmap.getHeight() / 2,
                    bitmap.getWidth() / 2, paint);
            paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
            canvas.drawBitmap(bitmap, rect, rect, paint);
            return output;
        }
    }
}

File: app/src/main/java/com/example/spotibae/Adapter/Adapter.kt
package com.example.spotibae.Adapter

import com.example.spotibae.Models.ModelClass
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import android.view.LayoutInflater
import com.example.spotibae.R
import com.example.spotibae.Adapter.Adapter.ViewHolder.MyClickListener
import android.content.Intent
import android.graphics.*
import com.example.spotibae.Activities.Messaging.MessagingScreen
import com.example.spotibae.Activities.Messaging.MusicPlayer
import android.widget.TextView
import com.google.firebase.storage.FirebaseStorage
import android.view.View
import android.widget.ImageView

class Adapter(private val userList: List<ModelClass>) : RecyclerView.Adapter<Adapter.ViewHolder>() {
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.fragment_chat_user, parent, false)
        return ViewHolder(view, object : MyClickListener {
            override fun onMessages(p: Int) {
                val intent = Intent(view.context, MessagingScreen::class.java)
                intent.putExtra("PROFILE_PIC_EMAIL", userList[p].email)
                intent.putExtra("receiveruid", userList[p].firebaseId)
                intent.putExtra("name", userList[p].textview1)
                view.context.startActivity(intent)
            }

            override fun onMusic(p: Int) {
                val intent = Intent(view.context, MusicPlayer::class.java)
                intent.putExtra("PROFILE_PIC_EMAIL", userList[p].email)
                intent.putExtra("USER_FIREBASE_ID", userList[p].firebaseId)
                intent.putExtra("USER_NAME", userList[p].textview1)
                view.context.startActivity(intent)
            }
        })
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val firebaseID = userList[position].firebaseId
        val email = userList[position].email
        val name = userList[position].textview1
        val msg = userList[position].textview2
        val time = userList[position].textview3
        holder.setData(firebaseID, email, name, msg, time)
    }

    override fun getItemCount(): Int {
        return userList.size
    }

    //view holder class
    class ViewHolder(itemView: View, listener: MyClickListener) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        var listener: MyClickListener
        var imageView: ImageView
        private val textView: TextView
        private val textView2: TextView
        private val textview3: TextView
        var music: ImageView
        var messages: ImageView
        fun setData(
            firebaseId: String?,
            email: String?,
            name: String?,
            msg: String?,
            time: String?
        ) {
            //imageView.setImageResource(resource);
            setImage(email)
            textView.text = name
            textView2.text = msg
            textview3.text = time
        }

        override fun onClick(view: View) {
            when (view.id) {
                R.id.messagesButton -> listener.onMessages(this.layoutPosition)
                R.id.musicPlayerButton -> listener.onMusic(this.layoutPosition)
                else -> {
                }
            }
        }

        interface MyClickListener {
            fun onMessages(p: Int)
            fun onMusic(p: Int)
        }

        fun setImage(email: String?) {
            val storageReference = FirebaseStorage.getInstance().reference
            val photoReference =
                storageReference.child("User").child(email!!).child("profilePic.png")
            val ONE_MEGABYTE = (1024 * 1024).toLong()
            photoReference.getBytes(ONE_MEGABYTE).addOnSuccessListener { bytes ->
                val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                imageView.setImageBitmap(Bitmap.createScaledBitmap(bitmap, 90, 90, true))
            }.addOnFailureListener { }
        }

        init {
            //here use xml ids
            //give different name not like constructor
            imageView = itemView.findViewById(R.id.imageview)
            textView = itemView.findViewById(R.id.textview)
            textView2 = itemView.findViewById(R.id.textview2)
            textview3 = itemView.findViewById(R.id.textview3)
            music = itemView.findViewById(R.id.musicPlayerButton)
            messages = itemView.findViewById(R.id.messagesButton)
            this.listener = listener
            music.setOnClickListener(this)
            messages.setOnClickListener(this)
        }
    }
}